### PR TITLE
implement odo describe for devfile

### LINKED ISSUE #3644: implement odo describe for devfile
`odo describe` should work with the devfile component including the JSON output.

The output format should be same for both types of components. 

/priority high
/area devfile

### PR DESCRIPTION
**Pull request by metacosm**
**What type of PR is this?**
/kind bug

**What does does this PR do / why we need it**:
Make `describe` work with devfiles.

**Which issue(s) this PR fixes**:

Fixes #3644 

**PR acceptance criteria**:

- [ ] Unit test 

- [ ] Integration test 

- [ ] Documentation 

- [ ] I have read the [test guidelines](https://github.com/openshift/odo/blob/master/docs/dev/test-architecture.adoc)

**How to test changes / Special notes to the reviewer**:

### CODE REVIEW

**Comment by girishramnani** [comment]
@metacosm can we please focus on implementing the `odo describe` feature and then open a follow up for cleanup as it is a time critical task

**Comment by metacosm** [comment]
> @metacosm can we please focus on implementing the `odo describe` feature and then open a follow up for cleanup as it is a time critical task

This is not a clean up. These are steps leading to the implementation as the code is a mess and I'm not about to duplicate the whole code that deals with s2i components just to get it to work on devfiles as well.

File: pkg/component/component_full_description.go
Comment (mik-dass) [inline pkg/component/component_full_description.go:133]: please handle the error here

**Comment by mik-dass** [comment]
Please add a integration test using `odo describe` for devfiles.

File: pkg/envinfo/envinfo.go
Comment (mik-dass) [inline pkg/envinfo/envinfo.go:137]: why not directly set the devfilePath in newEnvSpecificInfo() using envDir there?

Comment (metacosm) [reply pkg/envinfo/envinfo.go:137]: Because then I would need to duplicate the logic in that function to compute the starting point.

Comment (mik-dass) [reply pkg/envinfo/envinfo.go:137]: I didn't get what you mean by starting point. In most cases, `envDir` is the context directory. We can just join the default devfilePath i.e `./devfile.yaml` to it. Or am I missing something?

Comment (metacosm) [reply pkg/envinfo/envinfo.go:137]: I may be missing something but the starting point from which the devfile path is computed depends on several things. Look at the code for this function and you'll see what I mean.

Comment (metacosm) [reply pkg/envinfo/envinfo.go:137]: That said, I've seen different ways of dealing with this in different parts of the code so… 🤷

File: pkg/envinfo/envinfo.go
Comment (prietyc123) [inline pkg/envinfo/envinfo.go:295]: Shouldn't `EnvInfoFileExists` makes more sense than `Exists`

**Review by prietyc123 — CHANGES_REQUESTED** [review]

File: pkg/envinfo/envinfo.go
Comment (metacosm) [reply pkg/envinfo/envinfo.go:295]: No because it's part of the `LocalConfigProvider` interface which is also implemented by `LocalConfigInfo`.

Comment (kadel) [reply pkg/envinfo/envinfo.go:295]: This is correct. I don't like the long names that are duplicating package or struct names that we have in some places.

**Comment by metacosm** [comment]
1 test failing and not always the same one… seems like a flake to me.

**Comment by dgolovin** [comment]
@metacosm 'odo describe' reports component as not pushed, when it actually is

```
denis@denis-ThinkPad-P50:~/p/test-ws/nodejs-ex$ ~/go/src/github.com/openshift/odo/dist/bin/linux-amd64/odo list
Experimental mode is enabled, use at your own risk

Devfile Components: 
APP      NAME          PROJECT        TYPE       STATE
app1     nodejs-ex     project156     nodejs     Pushed

Openshift Components: 
APP      NAME           PROJECT        TYPE       SOURCETYPE     STATE
app1     component1     project156     nodejs     local          Pushed
denis@denis-ThinkPad-P50:~/p/test-ws/nodejs-ex$ ~/go/src/github.com/openshift/odo/dist/bin/linux-amd64/odo describe -o json
{
	"kind": "Component",
	"apiVersion": "odo.dev/v1alpha1",
	"metadata": {
		"name": "nodejs-ex",
		"namespace": "project156",
		"creationTimestamp": null
	},
	"spec": {
		"app": "app1",
		"type": "devfile",
		"urls": {
			"kind": "List",
			"apiVersion": "odo.dev/v1alpha1",
			"metadata": {},
			"items": null
		},
		"storages": {
			"kind": "List",
			"apiVersion": "odo.dev/v1alpha1",
			"metadata": {},
			"items": null
		},
		"ports": [
			"5858"
		]
	},
	"status": {
		"state": "Not Pushed"
	}
}
```

File: pkg/envinfo/envinfo.go
Comment (girishramnani) [inline pkg/envinfo/envinfo.go:137]: lets not do this

Comment (metacosm) [reply pkg/envinfo/envinfo.go:137]: Indeed, let's see if I can find a better way.

Comment (mik-dass) [reply pkg/envinfo/envinfo.go:137]: I checked the function again. You are right, my suggestion won't work in case of the env specified envDir.

File: pkg/component/component_full_description.go
Comment (johnmcollier) [inline pkg/component/component_full_description.go:60]: Return an error here rather than panicking

**Comment by johnmcollier** [comment]
So I can run `odo describe` on an unpushed devfile component, but it fails for me after pushing:
Before:
```
Johns-MacBook-Pro-3:node-devfile johncollier$ odo describe
Component Name: nodejs
Type: devfile: nodejs
```

After (on OpenShift):
```
Johns-MacBook-Pro-3:node-devfile johncollier$ odo describe
 ✗  deploymentconfigs.apps.openshift.io "nodejs-app" not found
```

After (on Kube):
```
Johns-MacBook-Pro-3:do-node johncollier$ odo describe
 ✗  the server could not find the requested resource (get deploymentconfigs.apps.openshift.io)
```

**Comment by metacosm** [comment]
> So I can run `odo describe` on an unpushed devfile component, but it fails for me after pushing:
> Before:
> 
> ```
> Johns-MacBook-Pro-3:node-devfile johncollier$ odo describe
> Component Name: nodejs
> Type: devfile: nodejs
> ```
> 
> After (on OpenShift):
> 
> ```
> Johns-MacBook-Pro-3:node-devfile johncollier$ odo describe
>  ✗  deploymentconfigs.apps.openshift.io "nodejs-app" not found
> ```
> 
> After (on Kube):
> 
> ```
> Johns-MacBook-Pro-3:do-node johncollier$ odo describe
>  ✗  the server could not find the requested resource (get deploymentconfigs.apps.openshift.io)
> ```

Yep, just realized this myself. The architecture is horrendous and makes it difficult to support all possible cases…

File: pkg/component/component.go
Comment (johnmcollier) [inline pkg/component/component.go:946]: Rather than panicking here, can you update `GetComponentFromDevfile` to also return an error value and return an error here?

File: pkg/occlient/occlient.go
Comment (johnmcollier) [inline pkg/occlient/occlient.go:3033]: 👍

File: tests/integration/devfile/cmd_devfile_describe_test.go
Comment (johnmcollier) [inline tests/integration/devfile/cmd_devfile_describe_test.go:1]: Can you add a rule to the makefile so that the tests in here can be run individually?

Something like
```
.PHONY: test-cmd-devfile-describe
test-cmd-devfile-describe:
	ginkgo $(GINKGO_FLAGS) -focus="odo devfile describe command tests" tests/integration/devfile/
```

File: tests/integration/devfile/cmd_devfile_describe_test.go
Comment (johnmcollier) [inline tests/integration/devfile/cmd_devfile_describe_test.go:44]: Can you add a test case for when it is pushed?

**Review by johnmcollier — CHANGES_REQUESTED** [review]
Left some review comments. Some good refactoring that I like too 👍 

Also, thoughts on changing the output slightly from:
```
Johns-MacBook-Pro-3:node-devfile johncollier$ odo describe
Component Name: mynode
Type: devfile: nodejs
```

To either:
```
Johns-MacBook-Pro-3:node-devfile johncollier$ odo describe
Component Name: mynode
Deployment mode: devfile
Component type: nodejs
```
or 
```
Johns-MacBook-Pro-3:node-devfile johncollier$ odo describe
Component Name: mynode
Type: nodejs (devfile)
```

**Comment by metacosm** [comment]
> Left some review comments. Some good refactoring that I like too 👍
> 
> Also, thoughts on changing the output slightly from:
> 
> ```
> Johns-MacBook-Pro-3:node-devfile johncollier$ odo describe
> Component Name: mynode
> Type: devfile: nodejs
> ```
> 
> To either:
> 
> ```
> Johns-MacBook-Pro-3:node-devfile johncollier$ odo describe
> Component Name: mynode
> Deployment mode: devfile
> Component type: nodejs
> ```
> 
> or
> 
> ```
> Johns-MacBook-Pro-3:node-devfile johncollier$ odo describe
> Component Name: mynode
> Type: nodejs (devfile)
> ```

I don't know about this point. I asked on Slack about it but didn't get an answer… /cc @kadel @girishramnani

**Comment by kadel** [comment]
> > Left some review comments. Some good refactoring that I like too 👍
> > Also, thoughts on changing the output slightly from:
> > ```
> > Johns-MacBook-Pro-3:node-devfile johncollier$ odo describe
> > Component Name: mynode
> > Type: devfile: nodejs
> > ```
> > 
> > 
> > To either:
> > ```
> > Johns-MacBook-Pro-3:node-devfile johncollier$ odo describe
> > Component Name: mynode
> > Deployment mode: devfile
> > Component type: nodejs
> > ```
> > 
> > 
> > or
> > ```
> > Johns-MacBook-Pro-3:node-devfile johncollier$ odo describe
> > Component Name: mynode
> > Type: nodejs (devfile)
> > ```
> 
> I don't know about this point. I asked on Slack about it but didn't get an answer… /cc @kadel @girishramnani

Does the user need to know if it is a devfile component or not? I would consider completely skipping that information from the normal output. Otherwise I would go with  `Deployment mode: devfile` option

**Comment by metacosm** [comment]
> Does the user need to know if it is a devfile component or not? I would consider completely skipping that information from the normal output. Otherwise I would go with `Deployment mode: devfile` option

So just: `Type: nodejs` without mention of `devfile`?

**Comment by kadel** [comment]
> @kadel So just: `Type: nodejs` without mention of `devfile`?

yes. I would go with just that.

File: tests/integration/devfile/cmd_devfile_describe_test.go
Comment (metacosm) [reply tests/integration/devfile/cmd_devfile_describe_test.go:47]: Done.

File: tests/integration/devfile/cmd_devfile_describe_test.go
Comment (metacosm) [reply tests/integration/devfile/cmd_devfile_describe_test.go:1]: Done.

File: pkg/component/component.go
Comment (metacosm) [reply pkg/component/component.go:943]: The something are the parameters :smile:

File: pkg/component/component.go
Comment (mik-dass) [inline pkg/component/component.go:820]: Some comments regarding the function

File: pkg/component/pushed_component.go
Comment (mik-dass) [inline pkg/component/pushed_component.go:216]: can we run these two network operations in parallel?

File: pkg/component/pushed_component.go
Comment (mik-dass) [inline pkg/component/pushed_component.go:264]: can we run these client operations in parallel?

File: pkg/component/pushed_component.go
Comment (mik-dass) [inline pkg/component/pushed_component.go:211]: Any reason why we are returning a list instead of a map?

Comment (mik-dass) [reply pkg/component/pushed_component.go:211]: Please add some comment for the function

File: pkg/component/pushed_component.go
Comment (metacosm) [reply pkg/component/pushed_component.go:216]: We could but I'd rather not complicate things at this point.

File: pkg/component/pushed_component.go
Comment (metacosm) [reply pkg/component/pushed_component.go:211]: Why not return a map?

File: tests/integration/devfile/cmd_devfile_describe_test.go
Comment (mik-dass) [inline tests/integration/devfile/cmd_devfile_describe_test.go:33]: Let's not change the directory

File: pkg/component/pushed_component.go
Comment (mik-dass) [inline pkg/component/pushed_component.go:255]: Please add unit tests for this and some of the above functions

File: pkg/component/pushed_component.go
Comment (girishramnani) [reply pkg/component/pushed_component.go:211]: ordering becomes predictable

File: tests/integration/devfile/cmd_devfile_describe_test.go
Comment (metacosm) [reply tests/integration/devfile/cmd_devfile_describe_test.go:33]: I merely ported the existing test to devfile so sure, I could do that but then why was it allowed for the existing test?

File: pkg/component/pushed_component.go
Comment (metacosm) [reply pkg/component/pushed_component.go:255]: These should already be covered by the existing tests, I think as this function is only meant to hide where the component data comes from. The external behavior should already be tested.

File: pkg/component/pushed_component.go
Comment (mik-dass) [reply pkg/component/pushed_component.go:211]: Because this function is only called by `GetComponentNames()` which loops over the map keys to get the names. But we can loop over the list of components and use `<pushed-component>.GetName()` to get the name of each component too. Unless we use the keys to find individual items, a map usage feels weird to me as we are looping over all the items.

File: tests/integration/devfile/cmd_devfile_describe_test.go
Comment (mik-dass) [reply tests/integration/devfile/cmd_devfile_describe_test.go:33]: We currently have a issue to remove those https://github.com/openshift/odo/issues/3801.

File: pkg/component/pushed_component.go
Comment (mik-dass) [reply pkg/component/pushed_component.go:216]: But for s2i, it will always be one extra unnecessary call before the correct one.

**Comment by mik-dass** [comment]
```
[mrinaldas@localhost project]$ odo url create example --port 3000 --host com
 ✓  URL example created for component: nodejs

To apply the URL configuration changes, please use `odo push`
[mrinaldas@localhost project]$ odo describe Component Name: nodejs
Type: nodejs
URLs:
 · URL named example will be exposed via 5858
[mrinaldas@localhost project]$ odo storage create --path /data
 ✓  Added storage nodejs-uzoy to nodejs

Please use `odo push` command to make the storage accessible to the component
[mrinaldas@localhost project]$ odo describe Component Name: nodejs
Type: nodejs
Storage:
 · nodejs-uzoy of size 1Gi mounted to /data
URLs:
 · URL named example will be exposed via 5858
```
it seems to show the wrong port for non pushed components. After push, the port is displayed correctly.

**Comment by metacosm** [comment]
/retest


> ```
> [mrinaldas@localhost project]$ odo url create example --port 3000 --host com
>  ✓  URL example created for component: nodejs
> 
> To apply the URL configuration changes, please use `odo push`
> [mrinaldas@localhost project]$ odo describe Component Name: nodejs
> Type: nodejs
> URLs:
>  · URL named example will be exposed via 5858
> [mrinaldas@localhost project]$ odo storage create --path /data
>  ✓  Added storage nodejs-uzoy to nodejs
> 
> Please use `odo push` command to make the storage accessible to the component
> [mrinaldas@localhost project]$ odo describe Component Name: nodejs
> Type: nodejs
> Storage:
>  · nodejs-uzoy of size 1Gi mounted to /data
> URLs:
>  · URL named example will be exposed via 5858
> ```
> 
> it seems to show the wrong port for non pushed components. After push, the port is displayed correctly.
This behavior is due to https://github.com/metacosm/odo/blob/05e992c1512a42cfa085805e3a08c70961003779/pkg/component/component_full_description.go#L222-L236, which I didn't touch. I'm not sure why this is done this way, actually…

**Comment by mik-dass** [comment]
> @mik-dass This behavior is due to https://github.com/metacosm/odo/blob/05e992c1512a42cfa085805e3a08c70961003779/pkg/component/component_full_description.go#L222-L236, which I didn't touch. I'm not sure why this is done this way, actually… and I'm therefore not sure what the fix should be. I'm tired of changing something to fix an issue just to break something else.

I had a look again and it seems we are using the debugPort for the ports in case of the devfile https://github.com/openshift/odo/pull/3843/files#diff-692effa1b0b7a9ed8c1f7bb49a977b9aR952 and there is no `GetPorts()` in case of devfile. We need to use `devfile.ToRepresentation()` to get the ports from the devfile.

File: pkg/component/pushed_component.go
Comment (metacosm) [reply pkg/component/pushed_component.go:212]: Sure, but there is no intrinsic order here which returning a list would imply. `GetComponentNames` should indeed return a sorted list, which I will do.

**Review by johnmcollier — APPROVED** [review]
Changes look fine to me, don't see any glaring issues. Tested manually too on OCP and Kube, works great. 

Gonna approve this now if there's no concerns, let's get this one in 👍 

/lgtm

**Comment by kadel** [comment]
There is something horribly wrong with our setup scripts.
This is why tests are failing :-( 
```
	bitbucket.org/ww/goautoneg@v0.0.0-20120707110453-75cd24fc2f2c: reading https://api.bitbucket.org/2.0/repositories/ww/goautoneg?fields=scm: 404 Not Found

```
I'm looking into it

**Comment by kadel** [comment]
> There is something horribly wrong with our setup scripts.
> This is why tests are failing :-(
> 
> ```
> 	bitbucket.org/ww/goautoneg@v0.0.0-20120707110453-75cd24fc2f2c: reading https://api.bitbucket.org/2.0/repositories/ww/goautoneg?fields=scm: 404 Not Found
> ```
> 
> I'm looking into it

fix in https://github.com/openshift/odo/pull/3984

**Comment by dgolovin** [comment]
@kadel should kind value be different for s2i and devfile components? currently the only way to recognize devfile component is to check for devfile.yaml in local context.

**Comment by metacosm** [comment]
Could I get an LGTM back, please? 😄

**Comment by girishramnani** [comment]
@dgolovin we can do that in a follow up

**Pull request merged**
